contracts/ring.md

```diff
@@ -0,0 +1,340 @@
+i.e. the function name with the parenthesized list of parameter types.
+Parameter types are split by a single comma.
+
+The state is appended as an extra argument to the function according to the contracts state type.
+( *** subject to change *** )
+
+Return values are encoded in the same way as arguments.
+
+The arguments are encoded according to their types as follows:
+
+Arguments of types uint, address, and bool are encoded as a big endian 256-bit word (32 bytes).
+
+Arguments of other types are encoded as a pointer to the position of their place in the calldata
+with (the number of arguments *  32) subtracted.
+
+String arguments are encoded with a 32 byte length (number of bytes) and as few 256-bit words
+as needed padded on the right with 0.
+
+A list is a series of cons cells each cell two 256-bit words, where the first word is data
+encoded in the same way as arguments and the second argument is either an address or
+nil (encoded as a 256-bit word with all bits = 1).
+
+A tuple of size N is encoded in the same way as N arguments (functions
+arguments are in reality a tuple).
+
+Records are encoded as tuples.
+
+Maps are encoded as a list of tuples, (key, value). ***This is subject to change***
+
+The contract code start by getting the function "name" (hash):
+
+```
+PUSH1 0
+CALLDATALOAD
+```
+
+Then for each function in the contract it executes
+
+```
+DUP1
+PUSH32 'function hash'
+EQ
+JUMPI function address
+```
+
+The above could also be implemented as a search tree looking at one
+byte at the time of the function hash if that produces smaller
+code. The compiler could also choose to truncate the hash to the
+shortest unique prefix and shift the incoming hash down. E.g if there
+are only three functions in the contract and their hashes starts with
+0xA, 0xB nd 0xC respectively we could have this code:
+
+```
+PUSH1 248     ; bits to shift left (256 - (prefix length * 8))
+SHR           ; shift down to prefix length
+DUP           ; for next comparison
+PUSH1 0xA     ; get prefix of first fun
+EQ            ; compare
+JUMPI #fun0xA ; jump
+PUSH1 0xB
+EQ
+JUMPI #fun0xB
+JUMP #fun0xC
+```
+
+(Given that the contract invocation checks that the function call is to a valid address/hash before calling AEVM.)
+
+The shortest unique suffix could also be used, and the hash could be ANDed with the suffix length instead.
+
+
+Then each exported function starts with an exported entry point where it executes
+code to fetch the arguments to memory
+
+```
+JUMPDEST
+PUSH32 1 ; Skip the function name in the calldata
+CALLDATALOAD
+...
+PUSH32 Arity ; Skip other arguments
+CALLDATALOAD
+
+; Fetch data to memory
+PUSH1  0  ; Address 0 in mem
+PUSH   Arity + 1 ; # words to skip and
+DUP              ; Size = calldatasize - (arity+1)
+CALLDATASIZE  
+SWAP1
+SUB         
+CALLDATACOPY     ; copy from [arity+1..Size] to mem(0)
+
+JUMPDEST         ; for local calls
+                 ; code for function body
+```
```

Comment:
Wouldn't it be simpler (and safer) to do the entry point computation outside the byte code? Each contract would store a list of entry points, and when calling a contract we look up the entry point for the called function and start execution from there. That would save a lot of boiler plate in the contract byte code.

Of course, there's the question of Solidity compatibility.

Reply:
Yes, I have been thinking about this.
Solidity could call address 0 and would work as before.
There are only two questions:
1. How do we call a function in another contract? We would have to extend the call function in the VM with an address.
2. Who and when is the call address calculated? How does the caller get to know the call address, which is dependent on the bytecode. (in practice that is)

Reply:
If we want to keep Solidity compatibility, I guess we'd add a new call instruction with the extra argument. If we pass it the index of the entry point, rather than the address, we'd be less dependent on the actual byte code.

Reply:
Yes, and if we limit the number of functions to 256 we only need one byte. Function 256 could switch on an extra arguments if needed. OK?

Reply:
Sounds good to me.